spt/OrangeBox/scripts/tests/test.cpp

```diff
@@ -0,0 +1,252 @@
+		{
+			RunAllTests(folder, generating);
+			return;
+		}
+
+		ResetIteration();
+		try
+		{
+			currentTest = testName;
+			if (generating)
+			{
+				generatingData = true;
+			}
+			else
+			{
+				runningTest = true;
+				testItems = GetTestData(TestDataFile(testName));
+			}
+			
+			g_TASReader.ExecuteScript(testName);
+			lastTick = g_TASReader.GetCurrentScriptLength() - 1;
+		}
+		catch(const std::exception& ex)
+		{
+			successfulTest = false;
+			Msg("[TEST] Error loading test: %s\n", ex.what());
+		}
+		catch (...)
+		{
+			successfulTest = false;
+			Msg("[TEST] Uncaught exception while loading test\n");
+		}
+	}
+
+	bool Tester::RequiredFilesExist(const std::string & testName, bool generating)
+	{
+		if (generating)
+			return FileExists(ScriptFile(testName));
+		else
+			return FileExists(ScriptFile(testName)) && FileExists(TestDataFile(testName));
+	}
+
+	std::string Tester::TestDataFile(const std::string & testName)
+	{
+		return GetGameDir() + "\\" + testName + DATA_EXT;
+	}
+
+	std::string Tester::ScriptFile(const std::string & testName)
+	{
+		return GetGameDir() + "\\" + testName + SCRIPT_EXT;
+	}
+
+	std::string Tester::GetFolder(const std::string & testName)
+	{
+		return GetGameDir() + "\\" + testName;
+	}
+
+	void Tester::OnAfterFrames()
+	{
+		try
+		{
+			if (!runningTest && !generatingData)
+				return;
+
+			++afterFramesTick;
+			if (afterFramesTick >= lastTick)
+				TestDone();
+		}
+		catch (const std::exception& ex)
+		{
+			Msg("[TEST] Error in test iteration: %s\n", ex.what());
+		}
+		catch (...)
+		{
+			Msg("[TEST] Uncaught exception during test iteration\n");
+		}
+	}
+
+	void Tester::DataIteration()
+	{
+		try
+		{
+			if (runningTest)
+				TestIteration();
+			else if (generatingData)
+				GenerationIteration();
+			++dataTick;
+		}
+		catch (const std::exception& ex)
+		{
+			Msg("Error logging data for test: %s\n", ex.what());
+			successfulTest = false;
+		}
+		catch (...)
+		{
+			Msg("Uncaught exception while logging data\n");
+		}
+	}
+
+	void Tester::TestIteration()
+	{
+		while (currentTestItem < testItems.size() && testItems[currentTestItem].Tick <= dataTick)
+		{
+			auto& item = testItems[currentTestItem];
+
+			if (item.Tick < dataTick)
+				continue;
+
+			auto& tracker = trackers[item.TrackerNo];
+			auto result = tracker->Validate(item.Data);
+
+			if (!result.successful)
+			{
+				successfulTest = false;
+				Msg("[TEST] Tracker %s difference at tick %d: %s\n", tracker->TrackerName().c_str(), item.Tick, result.errorMsg.c_str());
+			}
+
+			++currentTestItem;
+		}
+	}
+
+	void Tester::GenerationIteration()
+	{
+		for (std::size_t i = 0; i < trackers.size(); ++i)
+		{
+			auto& tracker = trackers[i];
+			std::string data = tracker->GenerateTestData();
+			auto selfValidation = tracker->Validate(data);
+
+			if (!selfValidation.successful)
+			{
+				Msg("[TEST] Tracker %s failed to self-validate on tick %d : %s", tracker->TrackerName().c_str(), dataTick, selfValidation.errorMsg.c_str());
```

Comment:
I thought this was supposed to throw an exception? If not then what's the try-catch for